Support Alchemy 4.0

tvdeyen opened the pull request:
Builds on top of #12

afdev82 committed:
allow Alchemy CMS v4.0

afdev82 committed:
allow till AlchemyCMS 5

afdev82 committed:
increase version of pg_search

afdev82 committed:
remove joins in associations from Alchemy::Element

tvdeyen committed:
Test with Rails 5 and Alchemy 4

tvdeyen committed:
Update dummy app to Rails 5.0

tvdeyen committed:
Add sprockets manifest

tvdeyen committed:
Use factory_bot_rails

That still supports factory_girl

tvdeyen committed:
Add Alchemy 4.0 migrations

tvdeyen committed:
Test with Ruby 2.5

tvdeyen committed:
Use puma as local dev server

tvdeyen committed:
Update travis yml for postgres service

tvdeyen committed:
Cache gems during builds on TravisCI

tvdeyen committed:
Stick with sassc < 2.2

See https://github.com/sass/sassc-ruby/issues/146

dhiraj14 committed:
Add spec for public/non-public elements in search

This commit adds specs to check for presence or absence of
search result if the Alchemy::Element is public or not.
If the element is public and content is searchable, then
the result is displayed else, it is not.

dhiraj14 commented on lib/alchemy/pg_search/element_extension.rb:
```diff
@@ -12,21 +12,21 @@
 
   has_many :searchable_essence_richtexts,
-    -> { joins(:element).where(searchable: true, alchemy_elements: {public: true}) },
+    -> { where(searchable: true, alchemy_elements: {public: true}) },
```
@tvdeyen Based on this change, if we try to get searchable_essence_richtexts for any instance of Alchemy::Element, say element.searchable_essence_richtexts, it gives
```
*** ActiveRecord::StatementInvalid Exception: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "alchemy_elements"
LINE 1: ... AND (alchemy_essence_texts.searchable = true AND alchemy_el...
                                                             ^
: SELECT  "alchemy_essence_texts".* FROM "alchemy_essence_texts" INNER JOIN "alchemy_contents" ON "alchemy_essence_texts"."id" = "alchemy_contents"."essence_id" WHERE "alchemy_contents"."element_id" = $1 AND (alchemy_essence_texts.searchable = true AND alchemy_elements.public = true) AND "alchemy_contents"."essence_type" = $2 ORDER BY "alchemy_contents"."position" DESC, "alchemy_contents"."id" DESC LIMIT $3
```
Though, this doesn't impact the search because of the SQL query that gets generated via pg_search_scope :full_text_search.

Since, there are similar associations for `searchable_essence_richtexts` and `searchable_essence_pictures`, 
I tried updating the association to 
```
  has_many :searchable_essence_texts,
    -> { joins(:element).where("alchemy_essence_texts.searchable = true AND alchemy_elements.public = true") } ,
    class_name: 'Alchemy::EssenceText',
    source_type: 'Alchemy::EssenceText',
    through: :contents,
    source: :essence
``` 
And now it responds back for this and other given associations, as well as for the search query.
I had to add the query in where clause in form of string because otherwise it was not taking the correct
params and giving error. 
Please let me know your thoughts on the same. Thanks!

tvdeyen replied:
@dhiraj14 thanks for the research. If it works with the String in the `where` clause then let's use this. Alternatively we could add [`.references(:alchemy_elements)`](https://apidock.com/rails/v6.0.0/ActiveRecord/QueryMethods/references) to tell Rails to include that table in the query, but using the String method is just fine.

tvdeyen commented on spec/features/fulltext_search_feature_spec.rb:
```diff
@@ -132,6 +132,44 @@ module Alchemy
+            expect(page).to have_content('This is a secret caption.')
+          end
+        end
+      end
```
Thanks for adding specs!

dhiraj14 committed:
Update element_extension queries to use include rather than join

In the previous commits, to fix the issue on search for
searchable_contents, joins was removed. But because of this,
the associated models cannot be retrieved, since similar issue
is faced. So, for eg, it is not possible to use
element.searchable_essence_texts since, it would raise error.
To fix this, `element` has been included. This makes sure that
the table gets loaded and the fields are available.
The other option was to use references(:alchemy_elements) but it
raised similar error since the table wasn't pre-loaded.
Also, the current search works as expected, so this change
doesn't impact the current working
